### DISCUSSION TITLE
How to setup unique accessibility IDs for Capacitor components to use with Appium

### DISCUSSION THREAD

**Discussion by rocklaam** [ionic-team/capacitor#5771] [Q&A]
Hey guys, my project has finished building an App using Capacitor and React. Currently, we are setting up a Java Appium project to automate test cases with this app. We are trying to add accessibility IDs or class names for React components like this:

![image](https://user-images.githubusercontent.com/29154789/179716510-ae61f7e3-19ec-4206-9ff5-7040a547fd6c.png)

However, when running the testing .apk (Android) and .ipa (iOS) file with the Appium inspector, all components were transformed into native flatform components like this: 
**[Android]**
![image](https://user-images.githubusercontent.com/29154789/179716889-56dab533-c276-47c1-9b4d-64a44ab3d47c.png)

**and [iOS]**
![image](https://user-images.githubusercontent.com/29154789/179716963-5f1a4919-f5b0-4f5e-8c09-3af561d88b67.png)

In Android, we can get component's ids via resource-id, but we can't in iOS builds.

Is our approach a proper solution or do we need to follow another way? In case our solution is correct, what do we need to get Ids for the iOS platform?

Looking forward to your answers, thank you very much!

**Comment by facundofacio** [comment]
@rocklaam have you found a solution? I am facing the same issue

**Comment by mlynch** [comment]
I'm confused. You're writing React in a web view? There's no reason you'd see native Android views for those elements. The way to add accessibility ids to elements in your web view in a capacitor app is to just add id attributes or something else and then use dom queries. This works very well when using webdriverio with appium

**Comment by facundofacio** [comment]
@mlynch am facing a similar [issue](https://github.com/ionic-team/capacitor/discussions/6568) with a Vue + Capacitor app

**Comment by mlynch** [comment]
@facundofacio exact same advice applies: just query elements by id or other DOM selectors. Here's an example from an e2e example we put together using webdriverio + appium: https://github.com/ionic-team/ionic-e2e-example/blob/main/tests/pageobjects/map.page.ts

Here's the official webdriverio docs for querying for elements: https://webdriver.io/docs/api/browser/$

I strongly recommend using webdriverio + appium rather than appium on its own

I'm really not sure what's going on in your example because it makes no sense why those would be recognized as native controls because they are in the webview. Make sure you are switching to the webview context as shown here: https://github.com/ionic-team/ionic-e2e-example/blob/fffed0d77fad002953ec108d2b40416e8d1e1766/tests/helpers/platform/index.ts#L21

**Comment by clorichel** [comment]
Facing the exact same issue: unable to safely target iOS elements with a (webdriver + appium) selector because the html element `id` isn't flowing down to the iOS component. `@capacitor/core@5.7.6`, `webdriverio@8.39.0` and BrowserStack users here.

@mlynch not sure how you got it working on your ionic-e2e-example, because we only see `contexts: [ 'NATIVE_APP' ]` when we do `getContexts`, so the `switchToWeb` call just fails. Appium Inspector agrees:

<img width="304" alt="Capture d’écran 2024-08-01 à 13 22 26" src="https://github.com/user-attachments/assets/da3f047f-bd6a-4116-bebd-636541e97145">

To be clear:
- we'd LOVE to use css selectors which would be cross-compatible (Android+iOS),
- and avoid having to rely on `$('//XCUIElementTypeButton[@name="whatever name"]')` selectors which would only be iOS compatible... But the webview context is just not available somehow 🤯 

@rocklaam @facundofacio did you guys get anywhere since then? 🤗

**Reply by facundofacio** [reply]
- Hi @clorichel In my case (Vuejs + Capacitor), I've found a workaround of the iOS tag issue through adding the tag to the `aria-label` attribute, instead of the `id`.
- And for the case of the different XML selectors depending on the mobile platform, I finally used a mapping dictionary like this:
```python
XMLTAGS_MAP = dict(
    android=dict(
        button="android.widget.Button",
        edit_text="android.widget.EditText",
        password_field="android.widget.EditText",
        other="android.view.View",
        view="android.view.View",
        link="android.view.View",
        text_view="android.widget.TextView",
        full_tree="*",
    ),
    ios=dict(
        button="XCUIElementTypeButton",
        edit_text="XCUIElementTypeTextField",
        password_field="XCUIElementTypeSecureTextField",
        other="XCUIElementTypeOther",
        view="XCUIElementTypeOther",
        link="XCUIElementTypeLink",
        text_view="XCUIElementTypeStaticText",
        full_tree="*",
    ),
)
```
